### PR TITLE
Clarify event loop integration and prioritization

### PR DESCRIPTION
**Pull request by shaseley**
- Reword priority spec section to more closely match the recently added meta explainer
 - Add a note in the HTML event loop processing section to describe the intent of the Scheduler priorities wrt event loop integration, along with an example strategy (which ~= what we do in Chromium)

***
<a href="https://pr-preview.s3.amazonaws.com/shaseley/scheduling-apis/pull/92.html" title="Last updated on Jun 5, 2024, 6:00 PM UTC (7395c3a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/scheduling-apis/92/0a23970...shaseley:7395c3a.html" title="Last updated on Jun 5, 2024, 6:00 PM UTC (7395c3a)">Diff</a>

### CODE REVIEW

File: spec/patches.md [hunk 1]
```diff
@@ -55,6 +55,26 @@ Modify step 2.1 to read:
...
+Note: the HTML specification enables per-[=task source=] prioritization by making the selection of
+the next [=task=]'s [=task queue=] in the event loop processing steps [=implementation-defined=].
+Similarly, this specification makes selecting between the next {{Scheduler}} task and the next task
+from an [=event loop=]'s [=task queues=] [=implementation-defined=], which provides UAs with the
+most scheduling flexibility.
+<br/><br/>
+But the intent of this specification is that the {{TaskPriority}} of {{Scheduler}} tasks would
+influence the event loop priority. Specifically, "{{TaskPriority/background}}" tasks and
+continuations are typically considered less important than most other event loop tasks, and
+"{{TaskPriority/user-blocking}}" tasks and "{{TaskPriority/user-visible}}" and
```
Comment (mmocny) [inline spec/patches.md:67]: Nit: "and user-blocking tasks and user-visible and user-blocking continuations" is hard to parse for a note.  I thought it was an error of needless repetition at first.

I think this is saying:
- user-blocking tasks
- user-visible continuations
- user-blocking continuations

I would re-word to something like:
"user-blocking tasks or continuations, as well as user-visible continuations (but not tasks), are typically..."

File: spec/scheduling-tasks.md [hunk 2]
```diff
@@ -12,15 +12,17 @@ This spec formalizes three priorities to support scheduling tasks:
   };
 </pre>
 
-<dfn enum-value for=TaskPriority>user-blocking</dfn> is the highest priority, and is meant to be
-used for tasks that are blocking the user's ability to interact with the page, such as rendering the
-core experience or responding to user input.
+<dfn enum-value for=TaskPriority>user-blocking</dfn> is the highest priority, and it is meant for
+tasks that should run as soon as possible, such that running them at a lower priority would degrade
+user experience. This could be (chunked) work that is directly in response to user input, or
```
Comment (mmocny) [inline spec/scheduling-tasks.md:17]: "work that is directly in response to user input" I would naively interpret as "needed before next paint", which this priority would not guarantee.

I know you are trying to imply this is important follow-up after initial feedback to user input ((via the "chunked" reference), but I would really consider being even more explicit about it, as only influencing "relatively likelihood", but that no task scheduling priority has any guarantees, or something.

It seems likely to me that folks would actually default assume "user-blocking" means render-blocking (next paint), especially in contrast to the naming of "user-visible" (if both are for visible input, what possible difference could there be?).

I also assume most folks (maybe not via reading spec, but via articles copying this list) won't read the algorithms just the descriptions.

File: spec/scheduling-tasks.md [hunk 3]
```diff
@@ -12,15 +12,17 @@ This spec formalizes three priorities to support scheduling tasks:
...
+updating the in-viewport UI state, for example.
 
-<dfn enum-value for=TaskPriority>user-visible</dfn> is the second highest priority, and is meant to
-be used for tasks that are observable to the user but not necessarily blocking user actions, such as
-updating secondary parts of the page. This is the default priority.
+<dfn enum-value for=TaskPriority>user-visible</dfn> is the second highest priority, and it is meant
+for tasks that will be visible to the user, but either not immediately or are not essential to user
```
Comment (mmocny) [inline spec/scheduling-tasks.md:21]: Just more of a continuation of the above comment... consider something like:

"...meant for tasks that will have useful side effects that are in some way observable by the user, but are not absolutely necessary for immediate feedback"

... just as a way to try and reduce confusion about the word "visible" implying visual feedback.

**Review by mmocny — APPROVED** [review]
Some suggests are attached, but the existing explanations are consistent and accurate, so I leave it to your judgement.

File: spec/scheduling-tasks.md (hunk 2 above)
Comment (shaseley) [reply spec/scheduling-tasks.md:17]: Thanks, good point -- and I forgot that I had added something like that to the meta-explainer based on other feedback. I added a paragraph to basically say user-blocking != render-blocking.

File: spec/scheduling-tasks.md (hunk 3 above)
Comment (shaseley) [reply spec/scheduling-tasks.md]: Changed.

File: spec/patches.md [hunk 4]
```diff
@@ -55,6 +55,26 @@ Modify step 2.1 to read:
...
     * |schedulerQueue|'s [=scheduler task queue/tasks=], if |schedulerQueue| is not null.
 
+Note: the HTML specification enables per-[=task source=] prioritization by making the selection of
+the next [=task=]'s [=task queue=] in the event loop processing steps [=implementation-defined=].
+Similarly, this specification makes selecting between the next {{Scheduler}} task and the next task
+from an [=event loop=]'s [=task queues=] [=implementation-defined=], which provides UAs with the
+most scheduling flexibility.
+<br/><br/>
+But the intent of this specification is that the {{TaskPriority}} of {{Scheduler}} tasks would
+influence the event loop priority. Specifically, "{{TaskPriority/background}}" tasks and
```
Comment (clelland) [inline spec/patches.md:65]: This sentence is still a bit hard to parse, with a number of "and"s and "as well as". Maybe a small change could help:

> Specifically, "background" tasks and continuations are typically considered less important than most other event loop tasks, _**while**_ "user-blocking" tasks and continuations, as well as "user-visible" continuations (but not tasks), are typically considered to be more important.

**Review by clelland — APPROVED** [review]

**Comment by shaseley** [comment]
Thanks @mmocny and @clelland!
